### PR TITLE
feat: login functionality for end users

### PR DESCRIPTION
**Pull request by MattLui**
To test login functionality: 

-  Send the following POST request to obtain the token for the default user: 

` curl -X POST \
  http://localhost:3000/rpc/login \
  -H 'Content-Type: application/json' \
  -d '{"email": "user@snowclone.com", "pass": "snowclone"}' `

- Use the JWT in the response to test making POST requests to the postgrest api.
- The default user's role changed to "member" because user is a reserved word in pgsql

### CODE REVIEW

**Review by timw-csi — CHANGES_REQUESTED** [review]
nice one, Matt, thank you.

the demonstration of login worked per your instructions. looks like apiSchema.sql has some unnecessary lines like 17 and 18.

and then this version re-introduces some commented out lines that were previously cleaned up like lines 35-36 and 53- end. there might be some more, will look more later.

File: sql/apiSchema.sql [hunk 1]
```diff
@@ -5,13 +5,17 @@ create role authenticator noinherit login password 'mysecretpassword';
 -- Create web anon role and grant usage on schema api, and then grant default read privileges
```
Comment (timw-csi) [inline sql/apiSchema.sql:5]: cut 5-7

File: sql/apiSchema.sql [hunk 2]
```diff
@@ -5,13 +5,17 @@ create role authenticator noinherit login password 'mysecretpassword';
...
-grant select on tables to web_anon;
+create role anon noinherit;
+grant anon to authenticator;
+grant usage on schema api to anon;
+
+create role member noinherit;
+grant member to authenticator;
+grant usage on schema api to member;
 
+alter default privileges in schema api
```
Comment (timw-csi) [inline sql/apiSchema.sql:17]: cut 17-18

File: sql/apiSchema.sql [hunk 3]
```diff
@@ -5,13 +5,17 @@ create role authenticator noinherit login password 'mysecretpassword';
...
 
-alter default privileges in schema api
-grant select on tables to web_anon;
+create role anon noinherit;
+grant anon to authenticator;
+grant usage on schema api to anon;
+
+create role member noinherit;
+grant member to authenticator;
+grant usage on schema api to member;
```
Comment (timw-csi) [inline sql/apiSchema.sql:16]: cut 17-18

File: sql/apiSchema.sql [hunk 4]
```diff
@@ -25,6 +29,9 @@ grant dev_admin to authenticator;
 alter default privileges in schema api grant all on TABLES TO dev_admin;
 alter default privileges in schema api grant all on SEQUENCES TO dev_admin;
 alter default privileges in schema api grant all on FUNCTIONS TO dev_admin;
+
+alter default privileges in schema api grant SELECT, UPDATE, INSERT, DELETE on TABLES TO member;
+alter default privileges in schema api grant SELECT, UPDATE on SEQUENCES TO member;
 -- grant all on all tables in schema api to dev_admin;
```
Comment (timw-csi) [inline sql/apiSchema.sql:35]: cut 35-36

File: sql/apiSchema.sql [hunk 5]
```diff
@@ -25,6 +29,9 @@ grant dev_admin to authenticator;
 alter default privileges in schema api grant all on TABLES TO dev_admin;
 alter default privileges in schema api grant all on SEQUENCES TO dev_admin;
 alter default privileges in schema api grant all on FUNCTIONS TO dev_admin;
+
+alter default privileges in schema api grant SELECT, UPDATE, INSERT, DELETE on TABLES TO member;
+alter default privileges in schema api grant SELECT, UPDATE on SEQUENCES TO member;
```
Comment (timw-csi) [inline sql/apiSchema.sql:34]: cut 35-36

File: sql/apiSchema.sql [hunk 6]
```diff
@@ -25,6 +29,9 @@ grant dev_admin to authenticator;
...
 -- grant all on all tables in schema api to dev_admin;
 -- grant all on all sequences in schema api to dev_admin;
```
Comment (timw-csi) [inline sql/apiSchema.sql:37]: cut 52- end of file

File: sql/apiSchema.sql [hunk 7]
```diff
@@ -5,13 +5,17 @@ create role authenticator noinherit login password 'mysecretpassword';
...
+create role anon noinherit;
+grant anon to authenticator;
+grant usage on schema api to anon;
+
+create role member noinherit;
+grant member to authenticator;
+grant usage on schema api to member;
 
+alter default privileges in schema api
+grant select on tables to member;
```
Comment (timw-csi) [inline sql/apiSchema.sql:19]: could cut comments on lines 20-21 and 26-28

**Comment by timw-csi** [comment]
That's a lot of individual review comments, apologies for the visual spamming, but hopefully it's helpful.

Also, it looks like this branch is 8 commits behind main, so I think it'd be best to merge main into this branch (get all the updates `logins` branch is missing from `main` ) before we merge this into main. 

Thanks for your work on this. Looking now into how we can avoid hardcoding the JWT secret.

**Comment by MattLui** [comment]
Thanks, Tim. Very helpful. I'll look into when I get home later this afternoon. 

This branch was checked out for a while as main got updated, good call on merging main into it, never thought to do that, will do. 

To avoid hard coding JWT I think should be simple for local version. Docker-compose can read from .env file that we ask our user to generate. 

For logins.sql, can use the method recommended in tutorial and ask user to upload sql with their secret.

**Pull request closed without merging**